src/core/config/npo/identifier.rs

```diff
@@ -1,96 +1,109 @@
+        Self(name)
+    }
+    pub fn as_str(self) -> &'a str {
+        self.0
+    }
+}
 impl Display for TypeName<'_> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{}", self.0)
     }
 }
 
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub struct FieldName<'a>(&'a str);
+impl<'a> FieldName<'a> {
+    pub fn new(name: &'a str) -> Self {
+        Self(name)
+    }
+    pub fn as_str(self) -> &'a str {
+        self.0
+    }
+}
 impl Display for FieldName<'_> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{}", self.0)
     }
 }
 
-struct FindFanOutContext<'a> {
+pub struct Identifier<'a> {
     config: &'a Config,
-    type_name: &'a str,
-    is_list: bool,
 }
 
-#[inline(always)]
-fn find_fan_out<'a>(
-    ctx: FindFanOutContext<'a>,
-    visited: &mut HashMap<TypeName<'a>, HashSet<FieldName<'a>>>,
-) -> HashMap<TypeName<'a>, HashSet<(FieldName<'a>, TypeName<'a>)>> {
-    let config = ctx.config;
-    let type_name: TypeName = TypeName(ctx.type_name);
-    let is_list = ctx.is_list;
-    let mut ans = HashMap::new();
-
-    if let Some(type_) = config.find_type(type_name.0) {
-        for (field_name, field) in type_.fields.iter() {
-            let cur: FieldName = FieldName(field_name.as_str());
-            let tuple: (FieldName, TypeName) = (cur, TypeName(field.type_of.as_str()));
-
-            let condition = visited
-                .get(&type_name)
-                .map(|v: &HashSet<FieldName>| v.contains(&cur))
-                .unwrap_or_default();
-
-            if condition {
-                continue;
-            } else {
-                visited.entry(type_name).or_default().insert(cur);
-            }
+impl<'a> Identifier<'a> {
+    pub fn new(config: &'a Config) -> Self {
+        Self { config }
+    }
 
-            if field.has_resolver() && !field.has_batched_resolver() && is_list {
-                ans.entry(type_name).or_insert(HashSet::new()).insert(tuple);
-            } else {
-                let next = find_fan_out(
-                    FindFanOutContext {
-                        config,
-                        type_name: &field.type_of,
-                        is_list: field.list || is_list,
-                    },
-                    visited,
-                );
-                for (k, v) in next {
-                    ans.entry(k).or_insert(HashSet::new()).extend(v);
+    pub fn identify(self) -> Output<'a> {
+        let mut visited = HashMap::new();
+        if let Some(query) = &self.config.schema.query {
+            let yield_inner = self.find_fan_out(query, false, &mut visited);
+            yield_inner.into_output(query)
+        } else {
+            Default::default()
+        }
+    }
+    #[inline(always)]
+    fn find_fan_out(
+        &self,
+        type_name: &'a str,
+        is_list: bool,
+        visited: &mut HashMap<TypeName<'a>, HashSet<FieldName<'a>>>,
```

Comment:
```suggestion
        visited: Output<'a>,
```
It's taking in an Output and returning an `Output`.

Reply:
it's not output

Reply:
Can we use `HashSet<(TypeName, FieldName)>` instead?

Reply:
we should not use a tuple in HashSet imo..

Reply:
assertion `left == right` failed
  left: Output { map: {TypeName("F1"): {(FieldName("f2"), TypeName("String"))}, TypeName("Query"): {(FieldName("f1"), TypeName("F1"))}}, root: "Query" }
 right: Output { map: {TypeName("F1"): {(FieldName("f1"), TypeName("F1")), (FieldName("f2"), TypeName("String"))}, TypeName("Query"): {(FieldName("f1"), TypeName("F1"))}}, root: "Query" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

also it leads to incorrect result sometimes